### ISSUE TITLE
Can I do something to improve performance?

### ISSUE THREAD

**Issue by ewfuentes** [jeremy-rifkin/libassert#43]
Hello! Thank you for for all of the effort you've put into this library.

I'm trying it out in a very simple example and I was surprised by how long it takes to run.

Example:
```
#include "assert/assert.hpp"

#include "gtest/gtest.h"

TEST(AssertTest, assert_test) {
    ASSERT_THROW(VERIFY(false, "Expected true to be false"), libassert::verification_failure);
}
```
yields:
```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AssertTest
[ RUN      ] AssertTest.assert_test
Verification failed at third_party/assert_test.cc:7: virtual void AssertTest_assert_test_Test::TestBody(): Expected true to be false
    VERIFY(false, ...);

Stack trace:
# 1 std::string::size() const
      at basic_string.h:1064
# 2 void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
      at gtest.cc:2598
# 3 testing::internal::UnitTestImpl::os_stack_trace_getter()
      at gtest.cc:6188
# 4 testing::TestInfo::Run()
      at gtest.cc:2855
# 5 testing::TestSuite::Run()
      at gtest.cc:3013
# 6 testing::internal::UnitTestImpl::RunAllTests()
      at gtest.cc:5870
# 7 bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
      at assert_test:?
# 8 testing::UnitTest::Run()
      at gtest.cc:5443
# 9 main
      at gmock_main.cc:70

[       OK ] AssertTest.assert_test (4084 ms)
[----------] 1 test from AssertTest (4084 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (4084 ms total)
[  PASSED  ] 1 test.
```

It is surprising that it takes around 4 seconds to run this test?  I tried running it through valgrind and I found that a lot of time is spent in highlighting and "analysis". I tried disabling color output, but it didn't really seem to change how long it takes to run.

For what it's worth, I'm trying to integrate this into a bazel project, so I may have screwed up the compilation flags, but I thought I'd check here as well.

**Comment by jeremy-rifkin** [comment]
Hey! Thanks for much for your interest in the library!

Assertion failures can be slow (during development have assumed assert failures are exceptional so it's ok to be slow), but I wouldn't expect them to be this slow. My guesses for performance issues would be the series of addr2line invocations used to get stack trace symbols or stuff related to the syntax analysis as you mention. I'll try to reproduce locally and I expect improving performance will definitely be possible.

**Comment by jeremy-rifkin** [comment]
As a quick test when I do something like this:
```diff
--- a/include/assert.hpp
+++ b/include/assert.hpp
@@ -4,6 +4,10 @@
 // Copyright (c) 2021-2022 Jeremy Rifkin under the MIT license
 // https://github.com/jeremy-rifkin/libassert

+#include <chrono>
+#include <iostream>
+#include <ratio>
+
 #include <cstdio>
 #include <cstring>
 #include <sstream>
@@ -988,7 +992,11 @@ namespace libassert::detail {
         // process_args needs to be called as soon as possible in case errno needs to be read
         const auto processed_args = process_args(args_strings, args...);
         const auto fatal = processed_args.fatality;
+        auto start = std::chrono::steady_clock::now();
         void* raw_trace = get_stacktrace_opaque();
+        auto end = std::chrono::steady_clock::now();
+        auto diff = end - start;
+        std::cout << std::chrono::duration<double, std::milli>(diff).count() << " ms" << std::endl;
         // generate header
         std::string output;
         binary_diagnostics_descriptor binary_diagnostics;
```

I get the following:
```
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from AssertTest
[ RUN      ] AssertTest.assert_test
2622.92 ms
Verification failed at test.cpp:6: virtual void AssertTest_assert_test_Test::TestBody(): Expected true to be false
    VERIFY(false, ...);

Stack trace:
#  1 test.cpp    6 AssertTest_assert_test_Test::TestBody()
#  2 a.out       ? void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)
#  3 a.out       ? testing::Test::Run()
#  4 a.out       ? testing::TestInfo::Run()
#  5 a.out       ? testing::TestSuite::Run()
#  6 a.out       ? testing::internal::UnitTestImpl::RunAllTests()
#  7 a.out       ? bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
#  8 a.out       ? testing::UnitTest::Run()
#  9 gtest.h  2491 RUN_ALL_TESTS()
# 10 test.cpp   12 main

[       OK ] AssertTest.assert_test (2626 ms)
[----------] 1 test from AssertTest (2626 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (2626 ms total)
[  PASSED  ] 1 test.
```

So almost the entire time is spent generating the stack trace. Though I put a lot of work into the stack trace bodge I have had in mind setting up a proper stack trace solution, maybe backward cpp or whatnot, so it's probably finally time to actually do that and hopefully it'll be much faster.

**Comment by jeremy-rifkin** [comment]
I'm working on pulling all the stack trace stuff into another lib (https://github.com/jeremy-rifkin/cpptrace) and will be integrating it into libassert soon, it should be faster :)

**Comment by jeremy-rifkin** [comment]
One workaround: Use clang++ / dwarf 5. Apparently it's hundreds of times faster for addr2line.

**Comment by jeremy-rifkin** [comment]
This should now be resolved with #50, sorry it took so long to get a better solution in place!
